### PR TITLE
Added support for decoding OpenSSL RSA public keys in PEM format.

### PR DESCRIPTION
**Pull request by LamarLugli**
I added support for loading OpenSSL RSA public keys in PEM format.

You can generate the public key with OpenSSL using: 
`openssl rsa -in private.key -out public.key -pubout`

I also bumped up the version number, so it is ready to publish a nuget package.

### CODE REVIEW

File: examples/ConsoleAppNetCore/Demo.cs [hunk 1]
```diff
@@ -124,6 +125,51 @@ public static void TestPrivateKey()
...
+            var privateParameters = privateKeyCsp.ExportParameters(false);
+
+            if (!CompareByteArrays(publicParameters.Modulus, privateParameters.Modulus)
+                || !CompareByteArrays(publicParameters.Exponent, privateParameters.Exponent))
+            {
+                throw new Exception("public.key does not match private.key");
+            }
+        }
+
+        private static bool CompareByteArrays(byte[] arrayA, byte[] arrayB)
```
Comment (StefH) [inline examples/ConsoleAppNetCore/Demo.cs]: You can use https://msdn.microsoft.com/en-us/library/bb348567.aspx for this.

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.IO;
+
+namespace OpenSSL.PrivateKeyDecoder
+{
+    internal static class DecoderUtils
+    {
+        internal static bool CompareByteArrays(byte[] arrayA, byte[] arrayB)
+        {
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: Replace content from this method by https://msdn.microsoft.com/en-us/library/bb348567.aspx if possible.

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                }
+                i++;
+            }
+
+            return true;
+        }
+
+        internal static byte[] ExtractBytes(string text, string header, string footer)
+        {
+            var data = text;
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: string

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            data = data.Replace(header, string.Empty);
+            data = data.Replace(footer, string.Empty);
+
+            return Convert.FromBase64String(data);
+        }
+
+        internal static int GetFieldLength(BinaryReader reader)
+        {
+            var bt = reader.ReadByte();
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: byte bt

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+            //when 0x8? is specified, the value after the 8 is the number of bytes to read for the length
+            //we are going to assume up to 4 bytes since anything bigger is ridiculous, and 4 translates nicely to an integer
+
+            //.net is litte endian, whereas asn.1 is big endian, so just fill the array backwards.
+
+            var bytesToRead = bt & 0x0F;
+            var lengthArray = new byte[4];
+
+            for (var i = 4 - bytesToRead; i < 4; i++)
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: int i =

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                //when the value is less than 0x80, then it is a direct length.
+                return bt;
+            }
+
+            //when 0x8? is specified, the value after the 8 is the number of bytes to read for the length
+            //we are going to assume up to 4 bytes since anything bigger is ridiculous, and 4 translates nicely to an integer
+
+            //.net is litte endian, whereas asn.1 is big endian, so just fill the array backwards.
+
+            var bytesToRead = bt & 0x0F;
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: int bytesToRead =

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            {
+                lengthArray[4 - i - 1] = reader.ReadByte();
+            }
+
+            return BitConverter.ToInt32(lengthArray, 0);
+        }
+
+        internal static int GetIntegerSize(BinaryReader reader)
+        {
+            var bt = reader.ReadByte();
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: byte bt

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            if (bt != 0x02) // expect integer
+            {
+                return 0;
+            }
+
+            var count = GetFieldLength(reader);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: int count

File: src/OpenSSL.PrivateKeyDecoder/IOpenSSLPublicKeyDecoder.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    /// IOpenSSLPublicKeyDecoder
+    /// </summary>
+    public interface IOpenSSLPublicKeyDecoder
+    {
+        /// <summary>
+        /// Decode Public Key into a RSACryptoServiceProvider object. (Windows only)
+        /// </summary>
+        /// <param name="publicText">The public (rsa) key text.</param>
+        /// <returns>RSACryptoServiceProvider</returns>
+        RSACryptoServiceProvider Decode([NotNull] string publicText);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/IOpenSSLPublicKeyDecoder.cs:17]: Add [NotNullOrEmpty] ?

File: src/OpenSSL.PrivateKeyDecoder/IOpenSSLPublicKeyDecoder.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+        /// <summary>
+        /// Decode Public Key into a RSAParameters struct.
+        /// </summary>
+        /// <param name="publicText">The public text.</param>        
+        /// <returns>RSAParameters</returns>
+        RSAParameters DecodeParameters([NotNull] string publicText);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/IOpenSSLPublicKeyDecoder.cs:24]: Add [NotNullOrEmpty] ?

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        // 1.2.840.113549.1.1.1 - RSA encryption, including the sequence byte and terminal encoded null
+        private readonly byte[] OIDRSAEncryption = { 0x30, 0x0D, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01, 0x05, 0x00 };
+
+
+        private const string PublicKeyHeader = "-----BEGIN PUBLIC KEY-----";
+        private const string PublicKeyFooter = "-----END PUBLIC KEY-----";
+
+        /// <inheritdoc cref="IOpenSSLPublicKeyDecoder.Decode"/>
+        public RSACryptoServiceProvider Decode(string publicText)
+        {
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]:
``` cs
if (string.IsnNullOrEmpty(publicText))
{
    throw new ArgumentNullException(nameof(publicText));
}
```

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            // Create RSACryptoServiceProvider instance
+            var rsaCryptoServiceProvider = new RSACryptoServiceProvider();
+            rsaCryptoServiceProvider.ImportParameters(rsaParameters);
+            return rsaCryptoServiceProvider;
+        }
+
+        /// <inheritdoc cref="IOpenSSLPublicKeyDecoder.DecodeParameters"/>
+        public RSAParameters DecodeParameters(string publicText)
+        {
+            if (publicText == null)
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: Also check for string.IsnNullOrEmpty?

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        private RSAParameters DecodePublicKey(byte[] pkcs8)
+        {
+            // read the asn.1 encoded SubjectPublicKeyInfo blob
+            var memoryStream = new MemoryStream(pkcs8);
+
+            using (var reader = new BinaryReader(memoryStream))
+            {
+                var dataType = reader.ReadByte();
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: byte dataType

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+                var fieldLength = DecoderUtils.GetFieldLength(reader);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: int fieldLength

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+                byte[] seq = reader.ReadBytes(15);
+                if (!DecoderUtils.CompareByteArrays(seq, OIDRSAEncryption)) // make sure Sequence for OID is correct
+                {
+                    return default(RSAParameters);
+                }
+
+                var bt = reader.ReadByte();
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: byte bt

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+                if (bt != 0x03) // expect a bit string
+                {
+                    return default(RSAParameters);
+                }
+
+                var bitStreamLength = DecoderUtils.GetFieldLength(reader);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: int bitStreamLength

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+                bt = reader.ReadByte();
+                if (bt != 0x00) // expect a zero for number of bits in the bitstring that are unused
+                {
+                    return default(RSAParameters);
+                }
+
+                var dateType = reader.ReadByte();
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: byte dateType

File: src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+                if (dateType != 0x30) // data read as little endian order (actual data order for Sequence is 30 81)
+                {
+                    return default(RSAParameters);
+                }
+
+                var sequenceLength = DecoderUtils.GetFieldLength(reader);
```
Comment (StefH) [inline src/OpenSSL.PrivateKeyDecoder/OpenSSLPublicKeyDecoder.cs]: int sequenceLength

**Review by StefH — CHANGES_REQUESTED** [review]
Thanks for this PR, can you please take a look at my comments ?

**Comment by StefH** [comment]
Thanks a lot for this PR, I've added some comments, can you take a look?

**Comment by LamarLugli** [comment]
what's the reason against using var parameters on variables?

**Comment by StefH** [comment]
Normally, I use only `var` for complex objects. And keep the type for primitives like : bool, int, byte, long, string and so on....

**Comment by StefH** [comment]
Normally, I use only `var` for complex objects. And keep the type for primitives like : bool, int, byte, long, string and so on....

File: src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs (hunk 2 above)
Comment (LamarLugli) [reply src/OpenSSL.PrivateKeyDecoder/DecoderUtils.cs]: We can't do this because the library supports .Net Framework 2.0 and Linq is not supported.

**Review by StefH — APPROVED** [review]

**Pull request merged**